[5.2] Improvements to Arr::sortRecursive

mzur opened the pull request:
Continued from #9600.

GrahamCampbell commented on src/Illuminate/Support/Arr.php:
```diff
@@ -273,6 +273,20 @@ public static function has($array, $key)
 
     /**
+     * Determines if an array is associative.
+     * An array is "associative" if it doesn't have sequential numerical keys starting with 0.
```
You need another blank line before this line.

mzur commented:
Well, seems like I've done everything wrong I could possibly *do* wrong :smile: Sorry for that; I hope everything is ok now.

taylorotwell commented:
So is this a BC breaking feature or a bug fix?

taylorotwell commented:
Typically bug fixes would be merged into 5.1, BC breaking features would go into master / 5.2... so trying to figure out what is the proper fit here.

mzur commented:
It's not a bugfix since there is no specification for what `Arr::sortRecursive` should do exactly (e.g. there were no tests for it). It's an improvement of the function because in some cases it showed rather unexpected behavior (see the [initial PR](https://github.com/laravel/framework/pull/9600#issue-94590052)). Since this may be BC breaking and introduces a new "feature" (`Arr::isAssoc`) I'd say master/5.2 is correct.

taylorotwell commented:
I guess I can't figure out why this isn't a bug fix. The method description says recursively sort an array. The current method apparently doesn't?

lukasgeiter commented:
The fact that the array is now passed by reference could break things for some people.

mzur commented:
Sure, you could argue that it's a bugfix. But then we need to figure out how to make it non-BC breaking. I couldn't figure out a way to make this function work recursively without making the argument passed by reference. I reckon it can be done with a closure but that seems rather ugly to me.

lukasgeiter commented:
Can't you just assign the value back to the variable?

    $value = self::sortRecursive($value);

mzur committed:
Improves Arr::sortRecursive.
Recursion will now continue if a deeply nested non-assiciative array is encountered.
Also implements Arr::isAssoc to better distinguish between associative
and non-associuative arrays (having sequential numerical keys).

mzur commented:
@lukasgeiter That's a neat idea, done! I thought I missed something.

@taylorotwell Now it should be safe and fully BC. We could reopen #9600 to put this into 5.1.

taylorotwell commented:
OK Send it back to 5.1 then I guess :smile:
